internal/blockchain/blockindex.go

```diff
@@ -625,14 +633,26 @@ func (bi *blockIndex) HaveBlock(hash *chainhash.Hash) bool {
 // as a key in the block index.
 func shortBlockKey(hash *chainhash.Hash) uint32 {
 	// Use the first bytes of the hash directly since it is the result of a hash
-	// function that produces a uniformly-random distribution.  It is also worth
-	// noting that the mining process reduces entropy by zeroing the bits at the
-	// other end of the array, but there would need to be effectively impossible
-	// to achieve hash rates exceeding ~2^215.77 hashes/sec (aka ~89.8 peta
-	// yotta yotta hashes/sec) in order to start zeroing out the bits used here.
-	// Not only is that impossible for all intents and purposes, the only effect
-	// would be reducing overall memory savings due to increased collisions
-	// among the shortened keys.
+	// function that produces a uniformly-random distribution.
+	//
+	// It is also worth noting that the choice of bytes was made based on the
+	// possibility of the block hash and proof of work hash being the same and
+	// considering that the mining process reduces entropy by zeroing the bits
+	// at the other end of the array.
+	//
+	// Since the block hash and proof of work hash are independent, the entropy
+	// concern may or may not be relevant dependening on whether or not the same
+	// hash function is used (or was ever used historically in the case of hash
+	// algorithm changes), but this choice of bytes covers both possibilities
+	// without any notable downside.
+	//
+	// For the case when the block hash and proof of work hash are the same,
+	// there would need to be effectively impossible to achieve hash rates
+	// exceeding ~2^215.77 hashes/sec (aka ~89.8 peta yotta yotta hashes/sec) in
+	// order to start zeroing out the bits used here.  Not only is that
+	// impossible for all intents and purposes, the only effect would be
+	// reducing overall memory savings due to increased collisions among the
+	// shortened keys.
```

Comment:
Yeah, I noticed this as well, but I figured it wasn't an issue because it doesn't gain anything (worst case is a bit more memory usage, but overall it'd still be quite a bit less than it would be without the short keys) and, as you noted, there is a (slight) disincentive in the associated mining time.

I also thought about tying it to the pow hash instead of the block hash, but that would either (a) require storing the pow hash along with every `blockNode` thereby entirely defeating any memory savings the short keys provide (actually slightly worse versus not doing the short keys at all due to overhead); or (b) calculating the pow hash on every lookup which is significantly more expensive in terms of CPU usage and node lookups happen super frequently and thus would overall be way worse in all cases.